CMakeLists.txt

```diff
@@ -0,0 +1,99 @@
+cmake_minimum_required(VERSION 3.16.5)
+
+set(CMAKE_DISABLE_SOURCE_CHANGES ON)
+set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)
+
+set(CMAKE_VERBOSE_MAKEFILE on)
+include(toolchain.cmake) # TODO(melvinw): this should not be included here. It should be configured by ide or by passing it to cmake
+
+# Toolchain must be included before project declaration
+project(MarsRoverFirmware)
+
+# Global Settings
+set(COMPILE_FLAGS
+        -MMD
+        -MP
+
+        -O3
+        -g
+
+        -Wall
+        -Wextra
+        -Wpedantic
+        -Wvla
+        -Werror
+
+        -Wno-unused-parameter
+        -Wno-missing-field-initializers
+
+        -ffunction-sections
+        -fdata-sections
+        -fomit-frame-pointer
+        -funsigned-char
+
+        -fno-exceptions
+        -fno-delete-null-pointer-checks
+
+        -fstack-protector-strong
+
+        $<$<COMPILE_LANGUAGE:ASM>:-xassembler-with-cpp>
+        $<$<COMPILE_LANGUAGE:CXX>:-fno-rtti>
+        $<$<COMPILE_LANGUAGE:C,CXX>:-include${CMAKE_SOURCE_DIR}/config/mbed_config.h>
+        $<$<COMPILE_LANGUAGE:C,CXX>:-include${CMAKE_SOURCE_DIR}/config/rover_config.h>
+        $<$<COMPILE_LANGUAGE:C,CXX>:-include${CMAKE_SOURCE_DIR}/config/can_config.h>
+        )
+add_compile_options(${PROCESSOR_FLAGS} ${COMPILE_FLAGS})
+
+set(CMAKE_C_STANDARD 11)
+set(CMAKE_CXX_STANDARD 20)
+
+set(CMAKE_C_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+# Include mbed-os
+include(mbed-os.cmake)
+
+# Include library CMakeLists
+add_subdirectory(lib/can)
+add_subdirectory(lib/controllers)
+add_subdirectory(lib/miscellaneous)
+add_subdirectory(lib/neopixel)
+add_subdirectory(lib/pid)
+add_subdirectory(lib/sensors)
+add_subdirectory(lib/servo)
+
+# TODO(wmmc88): check if its in supported configuration list and warn if not in list. This should be in wrapper makefile
+if (NOT DEFINED APP)
+    message(FATAL_ERROR "APP variable not set in CACHE or ENV")
+elseif (NOT EXISTS "${CMAKE_SOURCE_DIR}/apps/${APP}")
+    message(FATAL_ERROR "${APP} app does not exist in the apps folder")
+endif ()
+
+if (NOT DEFINED TARGET)
+    message(FATAL_ERROR "TARGET variable not set in CACHE or ENV")
+elseif (NOT EXISTS "${CMAKE_SOURCE_DIR}/targets/${TARGET}")
+    message(FATAL_ERROR "${TARGET} target does not exist in the targets folder")
+endif ()
+message(STATUS "Configuring and generating build for ${APP} app on ${TARGET} target")
+
+# Add CMakeList for target-specific files
+add_subdirectory(targets/${TARGET})
+
+# Include Rover App CMakeLists
+add_subdirectory(apps/arm)
+add_subdirectory(apps/gimbtonomy)
+add_subdirectory(apps/safety)
+add_subdirectory(apps/science)
+
+# Include Test App CMakeLists
```

Comment:
wildcarding in a build system is a bad thing to do. The philosophy of all modern build systems(ex. bazel, cmake) is to be as explicit as possible a.k.a. the build system only does exactly what you tell it to do, no less, no more. Modern projects using make also employ this by forcing you to list exactly what files you want to build.

One of the main reasons of this is so that bugs and other issues are not introduced by flags accidentally being added or excluded, the wrong definition of a function being linked, wrong objects being linked against, etc.